flow/consent_types.go

```diff
@@ -460,6 +461,61 @@ type LogoutResult struct {
 	FrontChannelLogoutURLs []string
 }
 
+// Contains information on an ongoing device grant request.
+//
+// swagger:model DeviceUserAuthRequest
+type DeviceUserAuthRequest struct {
+	// ID is the identifier ("device challenge") of the device grant request. It is used to
+	// identify the session.
+	//
+	// required: true
+	ID       string `json:"challenge"`
+	CSRF     string `json:"-"`
+	Verifier string `json:"-"`
+
+	// Client is the OAuth 2.0 Client that initiated the request.
+	Client   *client.Client `json:"-"`
+	ClientID string         `json:"client_id"`
+	// RequestURL is the original Device Authorization URL requested.
+	RequestURL string `json:"request_url"`
+
+	RequestedAt time.Time      `json:"-"`
+	HandledAt   sqlxx.NullTime `json:"handled_at"`
+	WasHandled  bool           `json:"-"`
+}
+
+// HandledDeviceUserAuthRequest is the request payload used to accept a device user_code.
+//
+// swagger:model verifyUserCodeRequest
+type HandledDeviceUserAuthRequest struct {
+	// ID is the identifier ("device challenge") of the device request. It is used to
+	// identify the session.
+	ID string `json:"challenge"`
+
+	Request *DeviceUserAuthRequest `json:"-" faker:"-"`
+	// RequestURL is the original Device Authorization URL requested.
+	RequestURL string `json:"request_url"`
+	// RequestedScope contains the OAuth 2.0 Scope requested by the OAuth 2.0 Client.
+	RequestedScope sqlxx.StringSliceJSONFormat `json:"requested_scope"`
+	// RequestedAudience contains the access token audience as requested by the OAuth 2.0 Client.
+	RequestedAudience sqlxx.StringSliceJSONFormat `json:"requested_access_token_audience"`
+
+	// Client is the OAuth 2.0 Client that initiated the request.
+	Client   *client.Client `json:"client"`
+	ClientID string         `json:"-"`
+
+	RequestedAt time.Time `json:"-"`
+
+	HandledAt  sqlxx.NullTime      `json:"handled_at"`
+	WasHandled bool                `json:"-"`
+	Error      *RequestDeniedError `json:"-"`
+}
+
+// HasError returns whether the request has errors.
+func (r *HandledDeviceUserAuthRequest) HasError() bool {
+	return r.Error.IsError()
```

Comment:
Good question, looks like [this](https://github.com/ory/hydra/blob/master/flow/consent_types.go#L86) handles the case of the error being nil. I suppose that this works because the method is not part of the struct, but part of the type.